### ISSUE TITLE
Kernel reconnect is not done correctly when the kernel is in busy state

### ISSUE THREAD

**Issue by skukhtichev** [ipython/ipykernel#270]
There is an issue while reconnecting to the kernel when the kernel is in `busy` state: kernel stops responding to the code execution requests if kernel reconnect happened during code execution. I opened the issue in the notebook project some time ago: https://github.com/jupyter/notebook/issues/2633. I've done additional investigation and found that the issue could be related to the ZMQ connection handling (ROUTER - DEALER channels). The issue happens because shell ZMQ channel could not correctly handle notebook server disconnect when kernel is in busy state (ZMQ channel blocked). When the websocket is disconnected Notebook server finishes websocket connection and closes stream and socket connection on the notebooks server (DEALER connection), but the information about identity continues exist in the kernel ZMQ infrastructure. ZMQ is designed in the following way: when the client (DEALER) connects to the kernel (ROUTER) with the same identity and this identity exists on ROUTER (kernel) then the new connection is dropped. ZMQ supports additional option ZMQ_ROUTER_HANDOVER that allows to drop already existent connection on the ROUTER if the new connection requests comes. By default this option is disabled.

Enabling `ZMQ_ROUTER_HANDOVER ` option for all ROUTER sockets (shell, stdin, control) could solve the issue. 

How to reproduce the issue in the notebook:
1. Start a Python notebook and wait until kernel is ready (the issue does not depend on kernel, I chose python kernel for reproducing, but any kernel could be used for test case.)
2. Start code cell execution with the following code:
```
import time
time.sleep(20)
```
3. While the cell is executing click `kernel -> reconnect` (you should click `reconnect` while the kernel is in `busy` state)
4. Wait until cell execution will be finished (the result will be displayed in the cell output)
5. Add new code cell and execute the following code: `print(1)`
6. The cell will change state to `executing` (`[*]`), but the kernel will be shown as idle and no output will be displayed.
Expected behavior: Execution result `1` should be displayed in the cell output for code `print(1)`

**Comment by takluyver** [comment]
@minrk could you have a look at this one?

**Comment by minrk** [comment]
I've spent much of today digging into this one. Setting ROUTER_HANDOVER does work as a workaround, but isn't super satisfactory as all kernels will need to do the same to get a fix. Perhaps we should do this, though, as I think fixing this in the notebook isn't going to be doable (read on for why!).

To get all kernels to benefit, I want the fix to be in the notebook package. At first, I thought that making sure we close previous connections prior to opening new ones would be sufficient, but it turns out we [already do this](https://github.com/jupyter/notebook/blob/5.3.1/notebook/services/kernels/handlers.py#L252). After lots of digging, this led me to file https://github.com/zeromq/libzmq/issues/2892, as I'm pretty sure there's a bug in zmq's reconnect machinery since it's sensitive to what's going on in the main thread at connection time, which isn't supposed to happen.

It just so happens that the message buffering we recently added alleviates this for the common case because a single frontend reconnecting no longer corresponds to a new zmq connection (the latest connection is re-used), only a new websocket. So with 5.3, this bug requires either two concurrent frontends or disabling the output buffering to reproduce.

**Comment by minrk** [comment]
implemented router-handover in #300

**Comment by hanleilei** [comment]
I'm using the latest jupyterhub+ipykernel+notebook and try to setup a jupyterhub + docker for my colleagues, some of guys are working in other place, but this problem still happened occasionally, after dig into this one, I found the internet connection is not stable. zmq have 5 different channels, the shell channel can not handover.

I try to test in this way, run a 2 minutes simulation, make notebook disconnect with ipykernel manually and connect immediately, I found in the code below:
in kernelbase.py:
```python
        reply_content = self.do_execute(code, silent, store_history,
                                        user_expressions, allow_stdin)

        # Flush output before sending the reply.
        sys.stdout.flush()
        sys.stderr.flush()
        # FIXME: on rare occasions, the flush doesn't seem to make it to the
        # clients... This seems to mitigate the problem, but we definitely need
        # to better understand what's going on.
        if self._execute_sleep:
            time.sleep(self._execute_sleep)

        # Send the reply.
        reply_content = json_clean(reply_content)
        metadata = self.finish_metadata(parent, metadata, reply_content)

        reply_msg = self.session.send(stream, u'execute_reply',
                                      reply_content, parent, metadata=metadata,
                                      ident=ident)

```
if the connection is stable, all the simulation result will print in the browser, which were using the ippub channel, and then sent the 'execute_reply' with shell channel, this cell run were done. 

if I manually make  disconnect and connect again, all the simulation result will print in the browser, then stuck, there are not  'execute_reply' info,  the cell is asterisk marked, the later cell can not run.

I upgrade the environment from python 3.5, but not sure is there any other package need to be upgraded?

Great appreciate
